src/Orchard/Tasks/Locking/Services/DistributedLockService.cs

```diff
@@ -0,0 +1,153 @@
+using System;
+using System.Data;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Autofac;
+using Orchard.Data;
+using Orchard.Environment;
+using Orchard.Exceptions;
+using Orchard.Logging;
+using Orchard.Services;
+using Orchard.Tasks.Locking.Records;
+
+namespace Orchard.Tasks.Locking.Services {
+
+    public class DistributedLockService : IDistributedLockService {
+        private readonly IMachineNameProvider _machineNameProvider;
+        private readonly ILifetimeScope _lifetimeScope;
+        private readonly IClock _clock;
+        private readonly object _semaphore = new object();
```

Comment:
I see. Even when the lock is only being held for the duration of acquiring the lock I'm not sure it's good to have a common lock: acquiring a lock should be fast, but due to it involving a lot more than a Monitor.TryEnter() (creating a separate lifetime scope and issuing a DB roundtrip in the case of SQL-based locks) it can easily be orders of magnitude slower (acquiring a .NET lock is in the order of magnitude of nanoseconds, a DB call in itself is measurable in milliseconds when the DB is on a separate server).

Now of course without having a lot of demanding use-cases for distributed locks already (where I use distributed locking from Helpful Libraries this is nowhere a bottleneck) I can't say can be a big issue unless you have thousands of lock acquisitions per second but it would be the best not to build in bottlenecks in the first place. Actually the best would be if we wouldn't need a second layer of locking at all and just the distributed one would work alone well enough.